hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastObjectType.java

```diff
@@ -17,75 +17,129 @@
-        super(fields);
-        this.name = name;
-        this.digest = "HZ_OBJECT(" + name + ")";
-        this.nullable = true;
+    public HazelcastObjectType(String name, List<Field> fields) {
+        this(name, fields, true);
     }
 
-    public HazelcastObjectType(final String name, final List<Field> fields, final boolean nullable) {
-        super(fields);
+    public HazelcastObjectType(String name, List<Field> fields, boolean nullable) {
+        super(null);
         this.name = name;
-        this.digest = "HZ_OBJECT(" + name + ")";
         this.nullable = nullable;
+        this.fields = fields;
+        computeDigest();
     }
 
     public String getTypeName() {
         return name;
     }
 
     @Override
-    protected void generateTypeString(final StringBuilder sb, final boolean withDetail) {
-        sb.append("HZ_OBJECT(")
-                .append(name)
-                .append(")");
+    public boolean isNullable() {
+        return nullable;
     }
 
     @Override
     public SqlTypeName getSqlTypeName() {
         return SqlTypeName.OTHER;
     }
 
+    /**
+     * @implNote This is called by {@link #computeDigest} and {@link #toString}, with
+     * and without detail respectively. {@link #getFullTypeString}, {@link #equals} and
+     * {@link #hashCode} uses the {@link #digest} generated via this method.
+     * <p>
+     * A previous implementation used to ignore {@code withDetail} and only incorporated
+     * {@link #name} into the digest. This resulted in conflicts in {@link
+     * RelDataTypeFactoryImpl#DATATYPE_CACHE} in the test environment. However, the
+     * current state of user-defined types (UDTs) does not allow having different set of
+     * fields for the same UDT in different mappings. If this is allowed in the future,
+     * this implementation will ensure that there are no collisions in the cache.
+     */
     @Override
-    public @Nullable RelDataTypeField getField(final String fieldName, final boolean caseSensitive, final boolean elideRecord) {
-        assert fieldList != null;
-        RelDataTypeField fieldType = null;
-        for (final RelDataTypeField field : fieldList) {
-            if (fieldName.equals(field.getName())) {
-                fieldType = new RelDataTypeFieldImpl(fieldName, field.getIndex(), field.getType());
-                break;
+    protected void generateTypeString(StringBuilder sb, boolean withDetail) {
+        sb.append(name);
+        if (withDetail) {
+            sb.append('(');
+            for (Iterator<Field> it = fields.iterator(); it.hasNext();) {
+                RelDataTypeField field = it.next();
+                sb.append(field.getName()).append(':').append(field.getType().getFullTypeString());
+                if (it.hasNext()) {
+                    sb.append(", ");
+                }
             }
+            sb.append(')');
         }
+    }
 
-        return fieldType;
+    public void invalidate() {
```

Comment:
`Interner` (`RelDataTypeFactoryImpl.DATATYPE_CACHE`) javadoc says: 

> Warning: do not use with mutable objects.

Is that safe? `invalidate` changes `hashCode` so it can be called only before the object is interned. Is that the case here?

Reply:
`invalidate()` is only called when building the `HazelcastObjectType`.
1. For Hazelcast callers, field list _may_ need to be locked, but `HazelcastObjectType`s are not reused across queries or within the same query.
2. According to its contract, `RelDataType.getFieldList()` "returns read-only list of fields," so Calcite callers are not expected to change fields.

That's why `HazelcastObjectType` does not need to be invalidated during actual use.

Reply:
from the usage is seems safe, but IMO worth documenting and maybe even adding some assertions, eg. you can change fields when there is no digest yet and you cannot get digest until there are fields?